include/rtmidi17/detail/raw_alsa.hpp

```diff
@@ -0,0 +1,549 @@
+        msg.bytes.push_back(*bytes++);
+        runningStatusType_ = msg.bytes[0];
+      }
+
+      switch (message_type((uint8_t)runningStatusType_ & 0xF0))
+      {
+        case message_type::NOTE_OFF:
+        case message_type::NOTE_ON:
+        case message_type::POLY_PRESSURE:
+        case message_type::CONTROL_CHANGE:
+        case message_type::PITCH_BEND:
+          if(sz < 3)
+            return 0;
+
+          msg.bytes.push_back(*bytes++);
+          return 3;
+
+        case message_type::PROGRAM_CHANGE:
+        case message_type::AFTERTOUCH:
+          return 2;
+
+        default:
+          // TODO
+          return sz;
+      }
+    }
+  }
+};
+
+class midi_in_raw_alsa final : public midi_in_default<midi_in_raw_alsa>
+{
+public:
+  static const constexpr auto backend = "Raw ALSA";
+
+  midi_in_raw_alsa(std::string_view clientName, unsigned int queueSizeLimit)
+    : midi_in_default<midi_in_raw_alsa>{nullptr, queueSizeLimit}
+  {
+  }
+
+  ~midi_in_raw_alsa() override
+  {
+    // Close a connection if it exists.
+    midi_in_raw_alsa::close_port();
+  }
+
+  rtmidi::API get_current_api() const noexcept override
+  {
+    return rtmidi::API::LINUX_ALSA_RAW;
+  }
+
+  void open_port(unsigned int portNumber, std::string_view) override
+  {
+    if (connected_)
+    {
+      warning("midi_in_raw_alsa::open_port: a valid connection already exists.");
+      return;
+    }
+
+    auto device_list = get_device_enumerator();
+    device_list.enumerate_cards();
+
+    unsigned int num = device_list.outputs.size();
+    if (portNumber >= num)
+    {
+      error<no_devices_found_error>("midi_in_raw_alsa::open_port: no MIDI output sources found.");
+      return;
+    }
+
+    const int mode = SND_RAWMIDI_NONBLOCK;
+    const char* portname = device_list.outputs[portNumber].device.c_str();
+    int status = snd_rawmidi_open(&midiport_, nullptr, portname, mode);
+    if (status < 0)
+    {
+      error<driver_error>("midi_in_raw_alsa::open_port: cannot open device.");
+      return;
+    }
+
+    init_pollfd();
+    switch(mode_)
+    {
+      case processing_mode::THREAD:
+      {
+        this->thread_ = std::thread{[this] {
+          running_ = true;
+          run_thread();
+        }};
+        break;
+      }
+      case processing_mode::MANUAL:
+      {
+        running_ = true;
+        break;
+      }
+    }
+
+    connected_ = true;
+  }
+
+  void init_pollfd()
+  {
+    int num_fds = snd_rawmidi_poll_descriptors_count(this->midiport_);
+
+    this->fds_.clear();
+    this->fds_.resize(num_fds);
+
+    snd_rawmidi_poll_descriptors(this->midiport_, fds_.data(), num_fds);
+  }
+
+  void run_thread()
+  {
+    static const constexpr std::chrono::milliseconds poll_timeout{50};
+
+    while(this->running_ && do_poll(poll_timeout))
+    {
+      ;
+    }
+  }
+
+  bool do_poll(std::chrono::milliseconds poll_timeout) override
+  {
+    // Poll
+    int err = ::poll(fds_.data(), fds_.size(), poll_timeout.count());
```

Comment:
This is a look up if new data is in a ring buffer. 
Unfortunately it is not time stamped. We do not know how old the data is. 

See
https://www.alsa-project.org/alsa-doc/alsa-lib/rawmidi.html